Reduce cs_main locks during ConnectTip/SyncWithWallets

jonasschnelli opened the pull request:
DisconnectTip and AcceptToMempool are still holding cs_main (not intended to improve this in this PR)

jonasschnelli committed:
Reduce cs_main locks during ConnectTip/SyncWithWallets

sipa commented on src/main.cpp:
```diff
@@ -2914,13 +2905,23 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
+            SyncWithWallets(tx, pindexNewTip, NULL);
+        }
+        // ... and about transactions that got confirmed:
+        BOOST_FOREACH(const CTransaction &tx, pblock->vtx) {
```
This is not correct, it needs to iterate through all the blocks that got connected, not just the new tip.

jonasschnelli commented:
As always when you want to improve the locking a little bit, the change gets quickly more invasive.

Update serval things:
* Fixed @sipa point (iterate though all connected blocks).
* Removed the `CBlock *` parameter from the `SyncTransaction` signal and added `CBlockIndex*` together with `int posInBlock`.
* Updated the wallet logic that it can handle the slightly different signal.

kazcw commented:
ActivateBestChainStep reuses a vector for ConnectTip's txConflicted to combine results from multiple blocks. ConnectTip passes txConflicted to mempoool.removeForBlock every time.

kazcw commented on src/main.cpp:
```diff
@@ -2914,13 +2910,21 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
+
         // When we reach this point, we switched to a new tip (stored in pindexNewTip).
 
         // Notifications/callbacks that can run without cs_main
```
Could put the new loops below this comment with the other non-cs_main notifications

kazcw commented on src/wallet/wallet.cpp:
```diff
@@ -1224,7 +1224,14 @@ int CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool fUpdate)
             ReadBlockFromDisk(block, pindex, Params().GetConsensus());
             BOOST_FOREACH(CTransaction& tx, block.vtx)
```
Why not just change this loop to be numeric and then use the `i`?

jonasschnelli replied:
Right. This loop is stupid at this point. The transaction always have a `posInBlock`. Will update.

kazcw commented on src/wallet/wallet.cpp:
```diff
@@ -3276,30 +3283,18 @@ CWalletKey::CWalletKey(int64_t nExpires)
-    if (mi == mapBlockIndex.end())
-        return 0;
-    const CBlockIndex* pindex = (*mi).second;
     if (!pindex || !chainActive.Contains(pindex))
```
Could remove the null-check; pindex has already been dereferenced

jonasschnelli commented:
@kazcw Thanks for the detailed code review!

Added a squashme commit with fixed for issues found by @kazcw.

kazcw commented:
You know what, I misread. `txConflicted` is an out-only parameter all the way down the call stack, so it was fine without the `txConflictedCurrent`. Sorry about that.

sipa commented:
Needs rebase.

sipa commented on src/main.cpp:
```diff
@@ -44,6 +44,7 @@
 #include <boost/filesystem/fstream.hpp>
 #include <boost/math/distributions/poisson.hpp>
 #include <boost/thread.hpp>
+#include <boost/tuple/tuple.hpp>
```
Use std::tuple instead?

sipa commented on src/main.cpp:
```diff
@@ -2723,19 +2724,14 @@ bool static ConnectTip(CValidationState& state, const CChainParams& chainparams,
-    list<CTransaction> txConflicted;
+    list<CTransaction> txConflictedCurrent;
     mempool.removeForBlock(pblock->vtx, pindexNew->nHeight, txConflicted, !IsInitialBlockDownload());
+    txConflicted.insert(txConflicted.end(), txConflictedCurrent.begin(), txConflictedCurrent.end());
```
This looks like it is not doing anything, as txConflictedCurrent will always be empty?

jonasschnelli replied:
Oh. Right. L2728 should make use of the `txConflictedCurrent`.
Fixed.

jonasschnelli commented:
Rebased, fixed sipas issue and switched to `std::tuple`

sipa commented on src/main.cpp:
```diff
@@ -2717,19 +2717,14 @@ bool static ConnectTip(CValidationState& state, const CChainParams& chainparams,
-    mempool.removeForBlock(pblock->vtx, pindexNew->nHeight, txConflicted, !IsInitialBlockDownload());
+    list<CTransaction> txConflictedCurrent;
+    mempool.removeForBlock(pblock->vtx, pindexNew->nHeight, txConflictedCurrent, !IsInitialBlockDownload());
+    txConflicted.insert(txConflicted.end(), txConflictedCurrent.begin(), txConflictedCurrent.end());
```
Why a temporary variable, and not just pass vchConflicted to mempool.removeForBlock()?

jonasschnelli replied:
Without a temporary variable, `mempool.removeForBlock` would always work though the combined list `txConflicted` that was initialized [here](https://github.com/bitcoin/bitcoin/pull/7946/files/9471f32a8fad340d0518ded6320fc1402f1d6286#diff-7ec3c68a81efff79b6ca22ac1f1eabbaR2917).

Although I'm not sure if this would be a problem, but at least it would slightly change currents master behavior.

sipa replied:
"Work through"? All that function does is append deleted CTransactions to it.

jonasschnelli replied:
Yes. Your right. Removed the temp. var.

sipa commented on src/main.cpp:
```diff
@@ -2919,6 +2912,8 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
         CBlockIndex *pindexNewTip = NULL;
         const CBlockIndex *pindexFork;
+        std::list<CTransaction> txConflicted;
+        std::vector<std::tuple<CTransaction,CBlockIndex*,int> > txChanged;
```
Sorry for changing my mind, but can you define a struct for this instead? Larger tuples don't really have good readability.

jonasschnelli replied:
IMO tuple is ideal for a such use case. It's only a temporary variable and its scope is very small. IMO an custom struct seems to be over complex.

sipa replied:
Fair enough.

laanwj replied:
c++ needs collections.namedtuple :)

sipa replied:
http://www.boost.org/doc/libs/1_58_0/libs/fusion/doc/html/fusion/container/map.html

dcousens replied:
given the repeated use,  maybe just `typedef` the `std::tuple`?  `tuple` has some nice conveniences over a `struct` alternative.

sipa commented on src/main.cpp:
```diff
@@ -2946,6 +2941,15 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
+        // throw all transactions though the signal-interface
+        // while _not_ holding the cs_main lock
+        BOOST_FOREACH(const CTransaction &tx, txConflicted)
+        SyncWithWallets(tx, pindexNewTip);
```
Indentation (and braces).

jonasschnelli replied:
Fixed Indentation and braces.

sipa commented on src/wallet/wallet.cpp:
```diff
@@ -3269,31 +3270,19 @@ CWalletKey::CWalletKey(int64_t nExpires)
-        return 0;
-    const CBlockIndex* pindex = (*mi).second;
-    if (!pindex || !chainActive.Contains(pindex))
+    if (!chainActive.Contains(pindex))
```
Access of chainActive without cs_main.

sipa commented on src/wallet/wallet.cpp:
```diff
@@ -912,11 +912,11 @@ void CWallet::MarkConflicted(const uint256& hashBlock, const uint256& hashTx)
-void CWallet::SyncTransaction(const CTransaction& tx, const CBlockIndex *pindex, const CBlock* pblock)
+void CWallet::SyncTransaction(const CTransaction& tx, const CBlockIndex *pindex, int posInBlock)
 {
     LOCK2(cs_main, cs_wallet);
```
This cs_main is unfortunate, but I guess we can get rid of it in a follow-up.

sipa commented:
utACK fdf8a5efe03004f4a581ba41b87737a4e1f1ff44

dcousens commented:
utACK fdf8a5e

laanwj commented:
utACK fdf8a5e, but the squashme commits need squashing

jonasschnelli commented:
Squashed.

sipa commented on src/wallet/wallet.cpp:
```diff
@@ -757,18 +757,18 @@ bool CWallet::AddToWallet(const CWalletTx& wtxIn, bool fFromLoadWallet, CWalletD
         AssertLockHeld(cs_wallet);
 
-        if (pblock) {
+        if (pIndex) {
```
Shouldn't this be `if (posInBlock != -1) {` instead? pindex is always passed, even when the transaction was removed.

jonasschnelli commented:
Squashed the squashme commit. Anything holding this back?

jtimon commented:
Decide on struct vs typedef of std::tuple<CTransaction,CBlockIndex*,int> (or just leave it as it is now, I guess)?
No strong opinion on that.
Fast review ACK 0d54074

laanwj commented:
Needs rebase.

jonasschnelli commented:
Rebased.
